Sources/Workspace/SourceControlPackageContainer.swift

```diff
@@ -141,6 +145,63 @@ internal final class SourceControlPackageContainer: PackageContainer, CustomStri
     public func getTag(for version: Version) -> String? {
         return try? self.knownVersions()[version]
     }
+    
+    /// Returns revision for the given tag.
+    public func getRevision(forVersion version: Version) throws -> Revision {
+        // FIXME: We need to get the revision here, and we don't have a
+        // way to get it back out of the resolver which is very
+        // annoying. Maybe we should make an SPI on the provider for this?
+        guard let tag = self.getTag(for: version) else {
+            throw InternalError("unable to get tag for \(self.package) \(version); available versions \(try self.versionsDescending())")
+        }
+        let revision = try self.getRevision(forTag: tag)
+        
+        try self.checkIntegrity(version: version, revision: revision)
```

Comment:
this change is very good, I wonder tho if we should make `checkIntegrity` public and call it explicitly from the workspace instead of making it part of `getRevision` which may be useful without the validation.

Reply:
`SourceControlPackageContainer` is `internal` though so we can make it that.

Reply:
https://github.com/apple/swift-package-manager/pull/3890/commits/0b17219c49abe088f703bb9c2e70571b9cb20a6c